### DISCUSSION TITLE
How to use `uv_fs_read` to read a single line from an input file?

### DISCUSSION THREAD

**Discussion by HallofFamer** [libuv/libuv#4393] [Q&A] [answered]
I am trying to use libuv to add asynchronous file IO for an experimental programming language, and I am a bit confused about how to use `uv_fs_read`. At this point I have figured out how to read to the end of the file, or read a certain number of characters into buffer. However, it seems that to read a single character(like `fget`) or a single line(like `fgets`) is not possible. The former is not very useful anyway, as it doesnt make much sense to read a single char asynchronously. But the latter is a functionality that Id like to implement in the PL, and I have no idea how or if it is even possible to do this. 

So here is my question, is it possible to use `uv_fs_read` to read a single line from an input file, if the position of the linebreak characters are not unknown ahead of time? If not, is there a workaround to this? Thanks.

**Comment by saghul** [comment, marked as answer]
You would need to read your til you find the line break.

Since uv_fs_read works with offsets, you can then resume reading where you left off.

**Reply by HallofFamer** [reply]
I see, so in this case do I need to just repeatedly call `uv_fs_read` to read 1 offset until the char it reads is a new line char `\n`? In this case, it will trigger the callback function each time it reads a char. Is there an easier way to trigger the callback only when new line char is read?

**Reply by saghul** [reply]
You could read bigger chunks and search for the line break in them.

**Reply by bnoordhuis** [reply]
The way fgets implements line reading is by reading into an internal buffer `s` and scanning said buffer for eol. Say you read `n` bytes, the output buffer is `d`, and disregarding bound checks:

1. if there's an eol at position `m`: `memcpy(d,s,m); memmove(s,s+m,n-m); n=read(fd,s+m,sizeof(s)-m);`
2. if there's no eol: `memcpy(d,s,n); n=read(fd,s,sizeof(s));`

Rinse & repeat.

**Reply by HallofFamer** [reply]
Thx for the suggestion, though in this case is there a way to know ahead of time how many chars/bytes are in the file so I can use a heuristic to determine the optimal buffer size? 

Also if I may ask, is there an equivalent function to fflush for Libuv?

**Reply by bnoordhuis** [reply]
You can `uv_fs_stat()` or `uv_fs_fstat()` but, since files can change between stat'ing and reading, only use it as a heuristic. "Special" files may report a size of zero (ex. files under /proc on linux.)

fflush is a libc/stdio thing, not an OS thing. The nearest equivalent is fsync/fdatasync, for which libuv has like-named wrappers.

**Reply by HallofFamer** [reply]
@bnoordhuis Thx for your suggestion, and yeah the buffer size is meant to be used as heuristics for calculating the default size. Regarding `fflush`, I guess I will just switch to use libuv equivalent version of `fsync` instead, I think they achieve almost the same thing. 

One more question though, if my understanding is correct, the `result` field from a `uv_fs_open` request contains the libuv file descriptor(`uv_file`) if file is opened successfully? In this case, I can also use `uv_get_osfhandle` to convert `uv_file` to `FILE*`, and `uv_open_osfhandle` to convert `FILE*` into `uv_file`?

**Reply by bnoordhuis** [reply]
No, or at least not directly, because it's not a FILE but a file descriptor (or a HANDLE on Windows.)

You can pass it to fdopen() however (_open_osfhandle + _fdopen on Windows, I think.)

**Reply by HallofFamer** [reply]
@bnoordhuis I see, thx for your answer. I have another question though, as I am aware of the fact that `uv_fs_read` accepts an offset as argument, and when -1 is supplied it uses the current offset. So is there a way to get the current offset value, kinda similar to how `ftell` works? Or is this not possible and we just have to keep track of the offset using a variable if needed?

**Reply by bnoordhuis** [reply]
Rule of thumb: always specify the position unless you're reading from a file where positional I/O doesn't work (like /dev files, pipes, append-only files, etc.)

Why: there's no telling what the actual file position is with concurrent readers/writers. That's also why there's no ftell/lseek equivalent in libuv.

**Reply by HallofFamer** [reply]
@bnoordhuis thx for your help and I am able to get libuv fs operations to work mostly on my code. On the other hand, I wonder what would be the recommended way to read a file to the end using libuv with callback? I heard you mentioned that using `uv_fs_stat’ may be able to give some idea of how much size is needed for the buffer, though there are caveats with this approach. What if the buffer size turns out not enough for reading till the end?

**Reply by bnoordhuis** [reply]
You read again (and again, and again) until you reach EOF. :-)

It's in general a good idea to take a stream-like approach and put some upper limit on the sizes of the chunks you read. You don't want to be slurping a 1 TB file all into memory.
